### ISSUE TITLE
Problem creating a nimbleFunction in another package

### ISSUE THREAD

**Issue by danielturek** [nimble-dev/nimble#93]
Might be trying to do this wrong, but I have a problem defining nimbleFunctions in other packages.  This code will reproduce the example, and the code underlying this package can be seen on Github.

`library(devtools)`
`install_github('danielturek/nimtest')`
`library(nimtest)`
`library(nimble)`
`nfDefinition`    ## finds it -- it created the NIMBLE function at build time
`nfDefinition()`  ## can't use it

Note, it also doesn't work if I include where = getLoadingNamespace() in the nimbleFunction definition.

**Comment by paciorek** [comment]
I think this can be addressed with this simple change. 

nfDefinition <- nimbleFunction(
  setup = function() {},
  run = function() print('ran nimble function ok'),
     where = topenv()
)

I think it may work to have topenv() as the default for 'where' in nimbleFunction() -- I just ran a few of the tests in test-mcmc, test-math, and test-user and things seem ok. If not, not sure if we want to set things up to automatically use topenv() when a user use nimbleFunction() in package code.

I won't claim to fully understand this, but when setRefClass is used to define a refObjectGenerator, the definition for the class (which seems to be an object with a name like ".__C__nfRefClass61" needs to be in the namespace of the package that is getting created. Since nimbleFunction() calls setRefClass, but  sets 'where' to be the global env, the definition is not being placed in the namespace. Setting 'where' to topenv() puts the definition in the package namespace during the package building process because topenv() gives back the package namespace in that situation.

**Comment by paciorek** [comment]
I will note that my only test of Daniel's package creation attempt was just to verify that nfDefinition() succeeds.

**Comment by danielturek** [comment]
Unfortunately, this doesn't appear to be as straightforward as we hoped.

I agree with Chris, that adding the argument "where = topenv()" to a nimbleFunction defined in the source of another package works.

However, changing (in NIMBLE package) the default argument value:
nimbleFunction <- function(......, where = topenv()) { .....}
did *not* work for me.  It failed when installing the other (dependent) package:
* installing to library ‘/Library/Frameworks/R.framework/Versions/3.2/Resources/library’
* installing *source* package ‘nimtest’ ...
** R
** preparing package for lazy loading
Error in assign(mname, def, where) : 
  cannot add bindings to a locked environment
Error : unable to load R code in package ‘nimtest’
ERROR: lazy loading failed for package ‘nimtest’
* removing ‘/Library/Frameworks/R.framework/Versions/3.2/Resources/library/nimtest’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.2/Resources/library/nimtest’

I also tried fidding with the default "where = " argument value, trying to make it eval in the right place, and find the correct environment, but I wasn't successful.

Chris, I'd be happy to chat with you about this.

**Comment by paciorek** [comment]
Hmm, I think what fails for Daniel does work for me. Daniel, let's look at
this together.

**Comment by danielturek** [comment]
Worked through this.  I found an appropriate default value for the `where` argument: essentially `topenv(parent.frame(3))`, to work its way *out* of the reference class environment hierarchy.  This default was moved into an auxillary function: `getNimbleFunctionEnvironment()`.  This appears to work fine for nimbleFunctions defined as package functions.

The functionality was extended to allow for package functions which themselves (internally) defined nimble functions.  This works, too, but gives some ugly looking warnings.  I'll open another issue for these warnings.

It's worth noting that the current implementation is entirely dependent upon our use of Reference Classes.  If we migrate away from these, then it would have to change.

These changes have been pushed to devel branch.

**Comment by paciorek** [comment]
I think the following may make the solution of topenv(parent.frame(3)) more robust.

Basic idea is to force evaluation of the default argument for 'where' in the frame of nimbleFunction(). This would allow us to use topenv(parent.frame()) and avoid the dependence on how the object (at the moment a refClass) is created.  I don't remember how to force evaluation but I think it's in Hadley's advanced R book. 

That said, if we move away from refClasses, then all of the 'where' business will presumably have to be reworked anyway. So not sure this is really making things more robust.

**Comment by perrydv** [comment]
?force

I think it’s worth making it work if we can even if it becomes moot later.

**Comment by paciorek** [comment]
force(where) within nimbleFunction() works, provided we use parent.frame(2) in Daniel's new getNimbleFunctionEnvironment().  Just a note - I have not changed devel.

This does get away from the dependence on the internals of how RefClasses work. The (2) just gets us to the frame that called nimbleFunction and is only reliant on Nimble's nested function calls.

**Comment by danielturek** [comment]
This sounds reasonable, Chris.  Yes, force() is the function to force
immediate argument evaluation.  I'm happy to test your changes if you push
your branch to the remote, or we can discuss on Monday.

Does your change prevent the warnings under the second test case?  (when a
package function internally defined a new NF)?

Daniel

**Comment by paciorek** [comment]
yes, it does work for a package fxn defining a new NF, either if the
package code then creates a NF instance or if the user in the Global loads
the package and creates the NF instance.

**Comment by paciorek** [comment]
branch on github is newRefClassLabels, in case you are interested.

**Comment by danielturek** [comment]
Chris's branch passed all my tests with no errors or warnings.

Daniel

**Comment by paciorek** [comment]
Ok, bringing this back from the dead...

In the nimble-users post from David Pleydell, I suggested he set

```
where = getNamespace('dependentPkgName')

```
and that worked. However, it is also the case that excluding the 'where' argument and defaulting to  getNimbleFunctionEnvironment() also works, at least in my simple test of his package. I believe we hashed things out correctly in this issue thread that I am reviving here. Namely,  getNimbleFunctionEnvironment() seems to properly use parent.frame to jump out of the nimbleFunction() call and then use topenv to make sure we are getting the namespace of a package or GlobalEnv. This seems like it is what we want when a user is defining a nimbleFunction either when writing code outside of or inside of a package.

Now it occurs to me that we may not need getLoadingNamespace for nimbleFunctions in the nimble package. It looks like getNimbleFunctionEnvironment() should return the nimble package namespace, just as getLoadingNamespace() does. I did a quick test with removing "where = getLoadingNamespace()" from "sampler_RW", rebuilding the package and running an MCMC that uses sampler_RW and things went fine.  I'll note that the need to include getLoadingNamespace was discussed in an email in June 2014, which preceded our development of getNimbleFunctionEnvironment() in this thread in September 2015. 

The one clear benefit of removing use of "where = getLoadingNamespace()" from nimbleFunctions defined in package code is that users who copy and modify our nimbleFunctions won't mistakenly keep using getLoadingNamespace and their code should just then work in their own packages. 

Even if we don't make that change, I recommend we:

1) followup once more on nimble-users and say that "where = getNamespace("dependentPkgName")" is not needed
2) add info to help(getLoadingNamespace) that makes it clear to users that user code should not use this.
